### PR TITLE
Validate the webhooks are registered

### PR DESCRIPTION
**Pull request by sthaha**
This patch changes the workflow as follows
1. The status after pipeline is installed is now set to `applied-pipeline`
   to avoid having to reapply the manifest if already applied successfully

2. Both deployment and mutation webhook is validate before setting the
   status to `validated-pipeline`. An additional delay of 15 seconds is
   applied before applying addons

This ensures that cluster-task defaults are set by the webhook when the
addons are applied.

Signed-off-by: Sunil Thaha <sthaha@redhat.com>

### CODE REVIEW

File: pkg/controller/config/config_controller.go [hunk 1]
```diff
@@ -292,42 +294,72 @@ func (r *ReconcileConfig) validatePipeline(req reconcile.Request, cfg *op.Config
...
+	if err != nil {
+		log.Error(err, "failed to validate mutating webhook")
+		_ = r.updateStatus(cfg, op.ConfigCondition{
+			Code:    op.PipelineValidateError,
+			Details: err.Error(),
+			Version: flag.TektonVersion})
+		return reconcile.Result{RequeueAfter: 15 * time.Second}, err
+	}
+	if !found {
+		return reconcile.Result{Requeue: true, RequeueAfter: 15 * time.Second}, nil
```
Comment (hrishin) [inline pkg/controller/config/config_controller.go]: What's the intent of requeing here again? :thinking:

File: pkg/controller/config/config_controller.go [hunk 2]
```diff
@@ -12,9 +12,9 @@ import (
 	"github.com/operator-framework/operator-sdk/pkg/predicate"
 	"github.com/prometheus/common/log"
 	op "github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
-	"github.com/tektoncd/operator/pkg/controller/transform"
-	"github.com/tektoncd/operator/pkg/controller/validate"
 	"github.com/tektoncd/operator/pkg/flag"
+	"github.com/tektoncd/operator/pkg/utils/transform"
+	"github.com/tektoncd/operator/pkg/utils/validate"
```
Comment (nikhil-thomas) [inline pkg/controller/config/config_controller.go]: :+1:

File: pkg/controller/config/config_controller.go [hunk 3]
```diff
@@ -292,42 +294,72 @@ func (r *ReconcileConfig) validatePipeline(req reconcile.Request, cfg *op.Config
...
-		return reconcile.Result{}, err
+		log.Error(err, "validating webhook deployment error")
+		return false, err
 	}
+
 	if !controller || !webhook {
-		log.Info("pipeline controller or/and pipeline webhook is not up")
-		return reconcile.Result{
-			Requeue:      true,
-			RequeueAfter: 5 * time.Second,
```
Comment (nikhil-thomas) [inline pkg/controller/config/config_controller.go]: @hrishin the above Reque{REqueafter: 15} is a replacement for this.
Now it is being separated for tekton-webhook and tekton-controller

File: pkg/controller/config/config_controller.go (hunk 2 above)
Comment (nikhil-thomas) [reply pkg/controller/config/config_controller.go]: @hrishin 
https://github.com/openshift/tektoncd-pipeline-operator/pull/143#discussion_r357475469

File: pkg/controller/config/config_controller.go (hunk 1 above)
Comment (nikhil-thomas) [reply pkg/controller/config/config_controller.go]: @hrishin https://github.com/openshift/tektoncd-pipeline-operator/pull/143#discussion_r357475469

Comment (nikhil-thomas) [reply pkg/controller/config/config_controller.go]: so if the deployment is not available yet, we need to check again, preferably a timed requeue than an a exponential requeue. So when tekton controller and tekton-webhook is available we can mark config condition `validated` and proceed with installing addons.

Comment (hrishin) [reply pkg/controller/config/config_controller.go]: Thanks  @nikhil-thomas for the explaining it :+1: . However wondering that change in state of the deployment would trigger reconciliation, thus do we need the requeue explicitely? 

Though nothing harmful as such on queueing it again, rather its guarantee that reconciliation loop would trigger again.

**Pull request merged**